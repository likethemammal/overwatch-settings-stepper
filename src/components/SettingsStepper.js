import React, { Component } from 'react'

import css from './SettingsStepper.less'

import _ from 'underscore'

import LeftSVG from 'svg-icon/dist/svg/mfglabs/chevron_left.svg'
import RightSVG from 'svg-icon/dist/svg/mfglabs/chevron_right.svg'

import general from '@likethemammal/general-components'

const { SVG } = general.components

import {
    activeOption as _activeOption,
    shouldShowLeftStepper as _shouldShowLeftStepper,
    shouldShowRightStepper as _shouldShowRightStepper,
} from '../selectors/stepper'

import {
    getNewActiveIndexFromStepLeft,
    getNewActiveIndexFromStepRight,
} from '../units/stepper/actions'

import * as validations from '../units/stepper/validations'

//todo: props warning if activeIndex is larger than options length

export default class StepperSwitch extends Component {

    static defaultProps = {
        options: [
            'On',
            'Off',
        ],
        activeIndex: 0,
    }

    onStepLeft = () => {
        const {
            activeIndex,
            onStep,
        } = this.props
        const shouldShowLeftStepper = _shouldShowLeftStepper(this.props)

        validations.NO_STEP_LEFT(shouldShowLeftStepper)

        onStep(
            getNewActiveIndexFromStepLeft(activeIndex)
        )
    }

    onStepRight = () => {
        const {
            options,
            activeIndex,
            onStep,
        } = this.props
        const shouldShowRightStepper = _shouldShowRightStepper(this.props)

        validations.NO_STEP_RIGHT(shouldShowRightStepper)

        onStep(
            getNewActiveIndexFromStepRight(activeIndex, options)
        )
    }

    render() {

        const {
            label,
        } = this.props

        const activeOption = _activeOption(this.props)
        const shouldShowLeftStepper = _shouldShowLeftStepper(this.props)
        const shouldShowRightStepper = _shouldShowRightStepper(this.props)

        const onMainStep = shouldShowRightStepper ? this.onStepRight : this.onStepLeft

        return <div className={css.container}>

            <div className={css.inner}>
                <div className={css.label}>
                    {label}
                </div>

                <div className={css.stepper}>
                    <button
                        className={css.left}
                        onClick={this.onStepLeft}
                        disabled={!shouldShowLeftStepper}
                    >
                        <SVG
                            raw={LeftSVG}
                        />
                    </button>
                    <div
                        className={css.active}
                        onClick={onMainStep}
                    >
                        {activeOption}
                    </div>
                    <button
                        className={css.right}
                        onClick={this.onStepRight}
                        disabled={!shouldShowRightStepper}
                    >
                        <SVG
                            raw={RightSVG}
                        />
                    </button>
                </div>
            </div>
        </div>
    }
}
