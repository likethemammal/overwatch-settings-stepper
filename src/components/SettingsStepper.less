@import "../../node_modules/@likethemammal/overwatch-general/lib/colors.less";

.container {
    padding: 3px 0 6px;

    &:hover {
        .inner {
            background: @gray_faded_setting_hover;
        }
    }

    * {
        box-sizing: border-box;
    }
}

.inner {
    composes: cf from '@likethemammal/general-components/bin/general.css';
    background: @black_faded_setting;
    transition: background 100ms ease-in;
    height: 40px;
    overflow: hidden;
}

.label {
    composes: __text from '@likethemammal/overwatch-general/lib/general.css';
    position: absolute;
    font-size: 24px;
    padding-left: 20px;
}

.stepper {
    composes: cf from '@likethemammal/general-components/bin/general.css';
    float: right;
}

.active {
    composes: __text from '@likethemammal/overwatch-general/lib/general.css';
    float: left;
    width: 300px;
    text-align: center;
    cursor: pointer;

    &:hover {
        background: @white_faded_row_hover;
    }
}

.icon_container {
    position: relative;
    width: 42px;
    padding: 5px 7px;
    float: left;
    vertical-align: middle;
    cursor: pointer;
    background: transparent;
    border: 0;

    &:disabled {
        opacity: 0.4;
        cursor: default;
    }

    &:hover:not([disabled]), &:focus:not([disabled]), &:active:not([disabled]) {
        background: @white_faded_row_hover;
        outline: 0;
    }

}

.left, .right {
    composes: icon_container;
}