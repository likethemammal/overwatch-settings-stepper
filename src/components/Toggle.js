import React, { Component } from 'react'

import SettingsStepper from './SettingsStepper'

import {
    activeIndex as _activeIndex
} from '../selectors/toggle'

import {
    getNewToggledFromStep
} from '../units/toggle/actions'

export default class Toggle extends Component {

    static defaultProps = {
        toggled: false,
        onToggle: null,
        label: ''
    }

    onStep = (activeIndex) => {

        const toggled = getNewToggledFromStep(activeIndex)

        this.props.onToggle(
            toggled
        )

    }

    render() {

        const {
            label,
        } = this.props

        const activeIndex = _activeIndex(this.props)

        return <SettingsStepper
            label={label}
            onStep={this.onStep}
            activeIndex={activeIndex}
        />
    }
}
