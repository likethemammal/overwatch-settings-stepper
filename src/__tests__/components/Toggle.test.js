import React from 'react'
import {mount} from 'enzyme'

import Toggle from '../../components/Toggle'
import SettingsStepper from '../../components/SettingsStepper'

import general from '@likethemammal/general-components'

const shared = general.__tests__._shared.tests

const label = 'some test label'
const toggled = true
const mockToggle = jest.fn()


const component_jsx = <Toggle
    label={label}
    toggled={toggled}
    onToggle={mockToggle}
/>

describe('Toggle', () => {
    describe('given a label, toggled value, and onToggle', () => {

        const component = mount(
            component_jsx
        )

        shared.SNAPSHOTS_SHOULD_MATCH([
            component
        ])

        test('SettingsStepper should have activeIndex of 0 if toggled is true', () => {
            expect(
                component.find(SettingsStepper).props().activeIndex
            ).toEqual(0)
        })

        test('SettingsStepper should receive label as label', () => {
            expect(
                component.find(SettingsStepper).props().label
            ).toEqual(label)
        })

        test('label should have label as text', () => {
            expect(
                component.find(SettingsStepper)
                    .childAt(0)
                    .childAt(0)
                    .childAt(0)
                    .instance().textContent
            ).toEqual(label)
        })

        test('onToggle is called when toggled is toggled', () => {
            component.props().onToggle({}, true)

            expect(
                mockToggle.mock.calls.length
            ).toEqual(1)
        })
    })
})