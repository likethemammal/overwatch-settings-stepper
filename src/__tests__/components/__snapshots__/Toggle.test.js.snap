// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle given a label, toggled value, and onToggle Snapshot 0 should match 1`] = `
<Toggle
  label="some test label"
  onToggle={[MockFunction]}
  toggled={true}
>
  <StepperSwitch
    activeIndex={0}
    label="some test label"
    onStep={[Function]}
    options={
      Array [
        "On",
        "Off",
      ]
    }
  >
    <div>
      <div>
        <div>
          some test label
        </div>
        <div>
          <button
            disabled={true}
            onClick={[Function]}
          >
            <t
              className="SVG-container"
              raw="test-file-stub"
            >
              <div
                className="SVG-container"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "test-file-stub",
                  }
                }
              />
            </t>
          </button>
          <div
            onClick={[Function]}
          >
            On
          </div>
          <button
            disabled={false}
            onClick={[Function]}
          >
            <t
              className="SVG-container"
              raw="test-file-stub"
            >
              <div
                className="SVG-container"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "test-file-stub",
                  }
                }
              />
            </t>
          </button>
        </div>
      </div>
    </div>
  </StepperSwitch>
</Toggle>
`;
